### PR TITLE
Fix dict object does not have isempty attribute error

### PR DESCRIPTION
**Pull request by alicertel**

### CODE REVIEW

File: drf_extra_fields/fields.py [hunk 1]
```diff
@@ -221,13 +221,23 @@ def to_representation(self, value):
         """
         Range instances -> dicts of primitive datatypes.
         """
-        if value.isempty:
-            return {'empty': True}
-        lower = self.child.to_representation(value.lower) if value.lower is not None else None
-        upper = self.child.to_representation(value.upper) if value.upper is not None else None
+        if isinstance(value, dict):
```
Comment (alicertel) [inline drf_extra_fields/fields.py:224]: If you provide a dict object to a serializer as `instance` the value is not converted to internal value.  (FooSerializer(data))

`data` of the serializer is run through validation.  In this case `value` is a range object. (FooSerializer(data=data))

**Comment by alicertel** [comment]
@omerfarukabaci @AtalayBaysal @gwtwod @vkhansen  I will test the updates on a project so this PR will be open for at least a week. In the mean time if you want you can install from the latest commit.

File: drf_extra_fields/fields.py (hunk 1 above)
Comment (omerfarukabaci) [reply drf_extra_fields/fields.py:224]: We may refactor this method as follows:

```python
def to_representation(self, value):
    """
    Range instances -> dicts of primitive datatypes.
    """
    if isinstance(value, dict) and not value:
        return value
    elif value.isempty:
        return {"empty": True}

    lower = value.get("lower") if isinstance(value, dict) else value.lower
    upper = value.get("upper") if isinstance(value, dict) else value.upper
    bounds = value.get("bounds") if isinstance(value, dict) else value._bounds

    return {"lower":  self.child.to_representation(lower) if lower is not None else None,
            "upper": self.child.to_representation(upper) if upper is not None else None,
            "bounds": bounds}
```

or 

```python
def to_representation(self, value):
    """
    Range instances -> dicts of primitive datatypes.
    """
    if isinstance(value, dict):
        if not value:
            return value

        lower = value.get("lower")
        upper = value.get("upper")
        bounds = value.get("bounds")
    else:
        if value.isempty:
            return {"empty": True}
        
        lower = value.lower
        upper = value.upper
        bounds = value._bounds

    return {"lower":  self.child.to_representation(lower) if lower is not None else None,
            "upper": self.child.to_representation(upper) if upper is not None else None,
            "bounds": bounds}
```

File: tests/test_fields.py [hunk 2]
```diff
@@ -479,7 +489,17 @@ class TestIntegerRangeField(FieldValues):
...
+        ({'lower': '1', 'upper': 2, 'bounds': '[)'},
+         {'lower': 1, 'upper': 2, 'bounds': '[)'}),
+        ({'lower': 1, 'upper': 2},
+         {'lower': 1, 'upper': 2, 'bounds': None}),
+        ({'lower': 1},
+         {'lower': 1, 'upper': None, 'bounds': None}),
+        ({'upper': 1},
+         {'lower': None, 'upper': 1, 'bounds': None}),
+        ({}, {}),
+    ]
```
Comment (omerfarukabaci) [inline tests/test_fields.py:502]: Should we add `NumericRange` instances here?

File: tests/test_fields.py [hunk 3]
```diff
@@ -520,6 +540,17 @@ def test_no_source_on_child(self):
...
+        ({'lower': Decimal('1.1'), 'upper': "2.3", 'bounds': '[)'},
+         {'lower': "1.1", 'upper': "2.3", 'bounds': '[)'}),
+        ({'lower': Decimal('1.1'), 'upper': "2.3"},
+         {'lower': "1.1", 'upper': "2.3", 'bounds': None}),
+        ({'lower': 1},
+         {'lower': "1", 'upper': None, 'bounds': None}),
+        ({'upper': 1},
+         {'lower': None, 'upper': "1", 'bounds': None}),
+        ({}, {}),
+    ]
```
Comment (omerfarukabaci) [inline tests/test_fields.py:553]: Should we add `NumericRange` instances here?

File: tests/test_fields.py [hunk 4]
```diff
@@ -560,6 +591,17 @@ class TestDecimalRangeFieldWithChildAttribute(FieldValues):
...
+        ({'lower': Decimal('1.1'), 'upper': "2.3", 'bounds': '[)'},
+         {'lower': "1.10", 'upper': "2.30", 'bounds': '[)'}),
+        ({'lower': Decimal('1.1'), 'upper': "2.3"},
+         {'lower': "1.10", 'upper': "2.30", 'bounds': None}),
+        ({'lower': 1},
+         {'lower': "1.00", 'upper': None, 'bounds': None}),
+        ({'upper': 1},
+         {'lower': None, 'upper': "1.00", 'bounds': None}),
+        ({}, {}),
+    ]
```
Comment (omerfarukabaci) [inline tests/test_fields.py:604]: Should we add `NumericRange` instances here?

File: tests/test_fields.py [hunk 5]
```diff
@@ -596,6 +638,17 @@ class TestFloatRangeField(FieldValues):
...
+        ({'lower': '1', 'upper': 2., 'bounds': '[)'},
+         {'lower': 1., 'upper': 2., 'bounds': '[)'}),
+        ({'lower': 1., 'upper': 2.},
+         {'lower': 1, 'upper': 2, 'bounds': None}),
+        ({'lower': 1},
+         {'lower': 1, 'upper': None, 'bounds': None}),
+        ({'upper': 1},
+         {'lower': None, 'upper': 1, 'bounds': None}),
+        ({}, {}),
+    ]
```
Comment (omerfarukabaci) [inline tests/test_fields.py:651]: Should we add `NumericRange` instances here?

File: tests/test_fields.py [hunk 6]
```diff
@@ -639,6 +692,25 @@ class TestDateTimeRangeField(TestCase, FieldValues):
...
+          'bounds': '[)'},
+         {'lower': '2001-01-01T13:00:00Z',
+          'upper': '2001-02-02T13:00:00Z',
+          'bounds': '[)'}),
+        ({'upper': '2001-02-02T13:00:00Z', 'bounds': '[)'},
+         {'lower': None, 'upper': '2001-02-02T13:00:00Z', 'bounds': '[)'}),
+        ({'lower': '2001-01-01T13:00:00Z', 'bounds': '[)'},
+         {'lower': '2001-01-01T13:00:00Z', 'upper': None, 'bounds': '[)'}),
+        ({}, {}),
+    ]
```
Comment (omerfarukabaci) [inline tests/test_fields.py:713]: Should we add `DateTimeTZRange` instances here?

File: tests/test_fields.py [hunk 7]
```diff
@@ -700,6 +772,25 @@ class TestDateRangeField(FieldValues):
...
+          'bounds': '[)'},
+         {'lower': '2001-01-01',
+          'upper': '2001-02-02',
+          'bounds': '[)'}),
+        ({'upper': '2001-02-02', 'bounds': '[)'},
+         {'lower': None, 'upper': '2001-02-02', 'bounds': '[)'}),
+        ({'lower': '2001-01-01', 'bounds': '[)'},
+         {'lower': '2001-01-01', 'upper': None,  'bounds': '[)'}),
+        ({}, {}),
+    ]
```
Comment (omerfarukabaci) [inline tests/test_fields.py:793]: Should we add `DateRange` instances here?

**Review by omerfarukabaci — CHANGES_REQUESTED** [review]
Thank you, looks great! I have made a few comments on tests and a refactoring suggestion.

File: tests/test_fields.py (hunk 2 above)
Comment (alicertel) [reply tests/test_fields.py:502]: No need, they are already being tested via test_outputs. I will move these to there as well.

File: drf_extra_fields/fields.py (hunk 1 above)
Comment (alicertel) [reply drf_extra_fields/fields.py:224]: Thanks, updated.

**Review by omerfarukabaci — APPROVED** [review]
Looks perfect, thank you! 🚀

**Pull request merged**